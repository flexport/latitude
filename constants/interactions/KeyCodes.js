/**
 * TEAM: growth
 * @flow strict
 */

export const DELETE = 46;
export const BACKSPACE = 8;
export const TAB = 9;
export const ENTER = 13;
export const ESC = 27;
export const LEFT = 37;
export const UP = 38;
export const RIGHT = 39;
export const DOWN = 40;
export const SHIFT = 16;
export const CAPS = 20;
export const COMMA = 188;
