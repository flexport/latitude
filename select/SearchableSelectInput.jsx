/**
 * TEAM: frontend_infra
 * @flow
 */
import * as React from "react";
import {StyleSheet, css} from "aphrodite";

import colors from "../colors";
import {type Size} from "../sizes";
import TextInput from "../TextInput";
import DropdownList, {type Option as DropdownOption} from "./DropdownList";
import useDropdown from "../tools/useDropdown";
import {defaultFilter} from "../TextInputAutocomplete";
import PopupWithClickAway from "../popup/PopupWithClickAway";
import {zIndices} from "../tools/zIndices";

import {BACKSPACE} from "../constants/interactions/KeyCodes";

export type Option<K> = {|
  ...DropdownOption,
  +value: K,
|};

type DropdownListProps = $Diff<
  React.ElementConfig<typeof DropdownList>,
  {|
    +options: any,
    +highlightedOption: any,
    +onClick: any,
    +isOpen: any,
  |}
>;

export type SelectInputProps<T> = {|
  // See DropdownList for dropdown props
  ...DropdownListProps,
  /** the current selected value of the select input */
  +value: T | null,
  /** the options the select input will display (see `Option` type for option's parameters) */
  +options: $ReadOnlyArray<Option<T>>,
  /** place holder text that will be displayed when the select input is empty */
  +placeholder?: string,
  /** whether the entire select input is disabled */
  +disabled?: boolean,
  /** the size of the input field */
  +size?: Size,
  /** whether the select input is in an invalid state */
  +isInvalid?: boolean,
  /** when true, this allows the user to select the empty element from the list. this calls onChange with `null`. */
  +isNullable?: boolean,
  /** whether the select input is in an prefilled state */
  +isPrefilled?: boolean,
  /** called when the value of the select input is changed */
  +onChange: (T | null) => void,
  /** called when the select input field is focused */
  +onFocus?: Event => void,
  /** called when the select input field is blurred */
  +onBlur?: Event => void,
  /** whether the popover generated uses a portal */
  +noPortal?: boolean,
|};

/**
 * @short Use SearchableSelectInput when constructing forms, if you need to select only one value from a list but the list is long.
 * @category Data Entry
 * @brandStatus V2
 * @status Stable
 */
function SearchableSelectInput<T>({
  sectionOrder,
  header,
  footer,
  value,
  options,
  placeholder = "",
  disabled = false,
  size = "m",
  isInvalid = false,
  isNullable = true,
  isPrefilled = false,
  onChange,
  onFocus,
  onBlur,
  noPortal = true,
}: SelectInputProps<T>) {
  const labelValueMap = options.reduce((acc, curr) => {
    acc.set(curr.label, curr.value);
    return acc;
  }, new Map());
  const valueLabelMap = options.reduce((acc, curr) => {
    acc.set(curr.value, curr.label);
    return acc;
  }, new Map());
  const [inputText, setInputText] = React.useState("");
  const [tabbingTextPlaceholder, setTabbingTextPlaceholder] = React.useState(
    valueLabelMap.get(value) || null
  );
  const popper = React.useRef();

  React.useEffect(() => {
    setTabbingTextPlaceholder(valueLabelMap.get(value));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [value]);

  const filteredDropdownOptions = options
    .filter(option => defaultFilter(inputText, option.label))
    .map(option => {
      const {value, ...dropdownOption} = option;
      return dropdownOption;
    });

  const handleChange = (newText: string | null) => {
    setInputText("");
    setTabbingTextPlaceholder(newText);

    onChange(labelValueMap.get(newText) || null);
  };

  const {highlightedIndex, handlers} = useDropdown(
    filteredDropdownOptions.map(option => option.label),
    {
      popper,
      onChange: handleChange,
      selectedLabel: valueLabelMap.get(value),
      rememberHighlightPosition: true,
      onFocus: (e: Event) => {
        if (onFocus) {
          onFocus(e);
        }
      },
      onBlur: (e: Event) => {
        setInputText("");
        setTabbingTextPlaceholder(valueLabelMap.get(value) || null);

        if (onBlur) {
          onBlur(e);
        }
      },
      onKeyDown: (e: KeyboardEvent) => {
        if (isNullable && e.keyCode === BACKSPACE && inputText === "") {
          handleChange(null);
          setTabbingTextPlaceholder("");
        }
      },
    }
  );

  const handleCancelClick = (e: Event) => {
    e.stopPropagation();

    handleChange(null);
    setInputText("");
  };

  const isInputTextPopulated =
    (tabbingTextPlaceholder && tabbingTextPlaceholder.length > 0) ||
    (inputText && inputText.length > 0);

  const suffix =
    isInputTextPopulated && isNullable
      ? {
          iconName: "cancel",
          onClick: handleCancelClick,
          size,
        }
      : {iconName: "downOpen"};

  return (
    <PopupWithClickAway ref={popper}>
      {(Target, Popup) => (
        <div
          className={css(
            styles.container,
            tabbingTextPlaceholder ? styles.darkPlaceholderText : null
          )}
        >
          <Target>
            <div className={css(styles.inputWrapper)}>
              <TextInput
                value={inputText}
                textOverflow="ellipsis"
                onChange={setInputText}
                onMouseDown={handlers.handleMouseDown}
                onFocus={handlers.handleFocus}
                onBlur={handlers.handleBlur}
                onKeyDown={handlers.handleKeyDown}
                isInvalid={isInvalid}
                disabled={disabled}
                placeholder={tabbingTextPlaceholder || placeholder}
                size={size}
                isPrefilled={isPrefilled}
                suffix={suffix}
              />
            </div>
          </Target>
          <Popup
            placement="bottom-start"
            zIndex={zIndices.zIndex1500AboveModal.value}
            noPortal={noPortal}
            className={noPortal ? css(styles.popup) : null}
          >
            <div
              tabIndex={-1}
              onFocus={handlers.handleFocus}
              onBlur={handlers.handleBlur}
              className={css(styles.dropdownContainer)}
            >
              <DropdownList
                options={filteredDropdownOptions}
                highlightedOption={
                  typeof highlightedIndex === "number"
                    ? filteredDropdownOptions[highlightedIndex].label
                    : null
                }
                sectionOrder={sectionOrder}
                header={header}
                footer={footer}
                onClick={handlers.handleItemClick}
              />
            </div>
          </Popup>
        </div>
      )}
    </PopupWithClickAway>
  );
}

const styles = StyleSheet.create({
  container: {
    position: "relative",
  },
  inputWrapper: {
    position: "relative",
  },
  darkPlaceholderText: {
    ":nth-child(1n) > div > div > div > input::placeholder": {
      color: colors.grey60,
    },
  },
  popup: {
    width: "100%",
  },
  dropdownContainer: {
    outline: "none",
    display: "block",
    position: "absolute",
    left: "0",
    top: "100%",
    padding: "8px 0",
    minWidth: "100%",
    zIndex: "10",
    transform: "none",
  },
});

export default SearchableSelectInput;
