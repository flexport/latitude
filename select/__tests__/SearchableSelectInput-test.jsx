/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {mount} from "enzyme";
import SearchableSelectInput from "../SearchableSelectInput";

import {UP, DOWN} from "../../constants/interactions/KeyCodes";

function mountSearchableSelectInput(props?: {}) {
  const options = [
    {label: "a", value: 1},
    {label: "b", value: 2},
    {label: "c", value: 3},
  ];

  return mount(
    <SearchableSelectInput
      options={options}
      value={1}
      onChange={() => {}}
      // $FlowFixMe(ctan) Flow issue from upgrade (1.111.3 => 1.115.0)
      {...props}
    />
  );
}

describe("SearchableSelectInput", () => {
  it("matches snapshot", () => {
    const comp = mountSearchableSelectInput();
    expect(comp).toMatchSnapshot();
  });

  it("disables", () => {
    const comp = mountSearchableSelectInput({disabled: true});
    expect(comp.find("input").props().disabled).toBe(true);
  });

  it("displays no text if value is null initially", () => {
    const comp = mountSearchableSelectInput({value: null});
    expect(comp.find("input").props().value).toBe("");
  });

  it("displays the label of the associated current value initially", () => {
    const comp = mountSearchableSelectInput({value: 3});
    expect(comp.find("input").props().placeholder).toBe("c");
  });

  it("should filter down on the number of options", () => {
    const comp = mountSearchableSelectInput({value: null});
    comp.find("input").simulate("change", {target: {value: "a"}});
    comp.find("input").simulate("focus");

    expect(comp.find("ul").children().length).toBe(1);
  });

  it("should filters options case insensitively", () => {
    const comp = mountSearchableSelectInput({value: null});
    comp.find("input").simulate("change", {target: {value: "B"}});
    comp.find("input").simulate("focus");

    expect(comp.find("ul").children().length).toBe(1);
  });

  it("disregards trailing whitespace when filtering", () => {
    const comp = mountSearchableSelectInput({value: null});
    comp.find("input").simulate("change", {target: {value: "     a"}});
    comp.find("input").simulate("focus");

    expect(comp.find("ul").children().length).toBe(1);
  });

  it("should open dropdown on focus", () => {
    const comp = mountSearchableSelectInput();
    expect(comp.find("DropdownList").length).toBe(0);
    comp.find("input").simulate("focus");
    expect(comp.find("DropdownList").length).toBe(1);
  });

  it("should open dropdown on mouse down", () => {
    const comp = mountSearchableSelectInput();
    expect(comp.find("DropdownList").length).toBe(0);
    comp.find("input").simulate("mousedown");
    expect(comp.find("DropdownList").length).toBe(1);
  });

  it("should close dropdown on mouse down if already open", () => {
    const comp = mountSearchableSelectInput();
    comp.find("input").simulate("focus");
    expect(comp.find("DropdownList").length).toBe(1);
    comp.find("input").simulate("mousedown");
    expect(comp.find("DropdownList").length).toBe(0);
  });

  it("should update the placeholder text when the value changes", () => {
    const comp = mountSearchableSelectInput();
    comp.setProps({value: 3});
    comp.update();
    expect(comp.find("input").props().placeholder).toBe("c");
  });

  describe("option highlighting", () => {
    it("remembers the last highlighted item on focus", () => {
      const comp = mountSearchableSelectInput({value: 2});
      comp.find("input").simulate("focus");
      const optionB = comp.find("DropdownOption").at(1);
      expect(optionB.props().isHighlighted).toBe(true);
    });

    it("remembers the last highlighted item on down press", () => {
      const comp = mountSearchableSelectInput({value: 2});
      comp.find("input").simulate("keyDown", {keyCode: UP});
      const optionB = comp.find("DropdownOption").at(1);
      expect(optionB.props().isHighlighted).toBe(true);
    });

    it("remembers the last highlighted item on up press", () => {
      const comp = mountSearchableSelectInput({value: 2});
      comp.find("input").simulate("keyDown", {keyCode: DOWN});
      const optionB = comp.find("DropdownOption").at(1);
      expect(optionB.props().isHighlighted).toBe(true);
    });
  });
});
