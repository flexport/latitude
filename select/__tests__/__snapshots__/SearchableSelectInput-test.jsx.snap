// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchableSelectInput matches snapshot 1`] = `
<SearchableSelectInput
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "a",
        "value": 1,
      },
      Object {
        "label": "b",
        "value": 2,
      },
      Object {
        "label": "c",
        "value": 3,
      },
    ]
  }
  value={1}
>
  <PopupWithClickAway
    escToClose={false}
    onClose={[Function]}
    onOpen={[Function]}
  >
    <Manager>
      <OnClickOutside(OnClickOutside)
        eventTypes={
          Array [
            "mousedown",
            "touchstart",
          ]
        }
        excludeScrollbar={false}
        onClickOutside={[Function]}
        outsideClickIgnoreClass="ignore-react-onclickoutside"
        preventDefault={false}
        stopPropagation={false}
      >
        <OnClickOutside
          disableOnClickOutside={[Function]}
          enableOnClickOutside={[Function]}
          eventTypes={
            Array [
              "mousedown",
              "touchstart",
            ]
          }
          onClickOutside={[Function]}
          outsideClickIgnoreClass="ignore-react-onclickoutside"
          preventDefault={false}
          stopPropagation={false}
        >
          <div
            className="container_e296pg-o_O-darkPlaceholderText_1sk0oij"
          >
            <Component>
              <Reference>
                <InnerReference
                  setReferenceNode={[Function]}
                >
                  <div
                    onKeyDown={[Function]}
                    role="presentation"
                  >
                    <div
                      className="inputWrapper_e296pg"
                    >
                      <TextInput
                        disabled={false}
                        isInvalid={false}
                        isPrefilled={false}
                        maxLength={255}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onMouseDown={[Function]}
                        placeholder="a"
                        readOnly={false}
                        size="m"
                        suffix={
                          Object {
                            "iconName": "cancel",
                            "onClick": [Function],
                            "size": "m",
                          }
                        }
                        textAlign="left"
                        textOverflow="ellipsis"
                        type="text"
                        value=""
                      >
                        <div
                          className="wrapper_oxngbz"
                        >
                          <input
                            className="baseNoPadding_656545-o_O-applyEllipsis_lspcml"
                            disabled={false}
                            maxLength={255}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            onKeyDown={[Function]}
                            onMouseDown={[Function]}
                            placeholder="a"
                            readOnly={false}
                            style={
                              Object {
                                "paddingLeft": 10,
                                "paddingRight": 32,
                                "textAlign": "left",
                              }
                            }
                            type="text"
                            value=""
                          />
                          <Sigil
                            disabled={false}
                            paddingDefaultTextInput={10}
                            side="right"
                            sigil={
                              Object {
                                "iconName": "cancel",
                                "onClick": [Function],
                                "size": "m",
                              }
                            }
                            size="m"
                          >
                            <span
                              className="sigil_1ogu4v7"
                              style={
                                Object {
                                  "right": "10px",
                                }
                              }
                            >
                              <IconButton
                                disabled={false}
                                height={
                                  Object {
                                    "type": "fixed",
                                  }
                                }
                                iconAlignment="left"
                                iconName="cancel"
                                intent="none"
                                isLoading={false}
                                kind="blank"
                                onClick={[Function]}
                                size="m"
                                type="button"
                                width="responsive"
                              >
                                <button
                                  className="base_i7qsly-o_O-m_12juzcr-o_O-blank_1tdfl6w-o_O-responsive_zk4wup-o_O-button_yfi27k-o_O-none_v41adi"
                                  disabled={false}
                                  onBlur={[Function]}
                                  onClick={[Function]}
                                  onFocus={[Function]}
                                  type="button"
                                >
                                  <Icon
                                    alignment="center"
                                    deprecatedAllowColorInheritance={true}
                                    iconName="cancel"
                                    isARIAAccessible={false}
                                    size="xs"
                                  >
                                    <span
                                      className="wrapper_17y0hv9 svgIcon"
                                      style={
                                        Object {
                                          "fill": "unset",
                                          "height": 12,
                                          "width": 12,
                                        }
                                      }
                                    >
                                      <svg
                                        aria-hidden="true"
                                        aria-labelledby="title desc"
                                        className="svgCenter_1r81nla"
                                        color="unset"
                                        height={12}
                                        role="img"
                                        style={
                                          Object {
                                            "height": "unset",
                                            "width": "unset",
                                          }
                                        }
                                        viewBox="0 0 64 64"
                                        width={12}
                                      >
                                        <title
                                          id="title"
                                        />
                                        <desc
                                          id="desc"
                                        >
                                          X
                                        </desc>
                                        <g
                                          fillRule="nonzero"
                                        >
                                          <path
                                            d="M53.213 46.142L39.071 32l14.142-14.142a1 1 0 0 0 0-1.414l-5.657-5.657a1 1 0 0 0-1.414 0L32 24.929 17.858 10.787a1 1 0 0 0-1.414 0l-5.657 5.657a1 1 0 0 0 0 1.414L24.929 32 10.787 46.142a1 1 0 0 0 0 1.414l5.657 5.657a1 1 0 0 0 1.414 0L32 39.071l14.142 14.142a1 1 0 0 0 1.414 0l5.657-5.657a1 1 0 0 0 0-1.414z"
                                            fillRule="evenodd"
                                            key="0"
                                            role="presentation"
                                          />
                                        </g>
                                      </svg>
                                    </span>
                                  </Icon>
                                </button>
                              </IconButton>
                            </span>
                          </Sigil>
                        </div>
                      </TextInput>
                    </div>
                  </div>
                </InnerReference>
              </Reference>
            </Component>
            <Component
              className="popup_zdxht7"
              noPortal={true}
              placement="bottom-start"
              zIndex={1500}
            >
              <div
                className="ignore-react-onclickoutside"
              >
                <Popper
                  placement="bottom-start"
                >
                  <InnerPopper
                    eventsEnabled={true}
                    placement="bottom-start"
                    positionFixed={false}
                    referenceElement={
                      <div
                        role="presentation"
                      >
                        <div
                          class="inputWrapper_e296pg"
                        >
                          <div
                            class="wrapper_oxngbz"
                          >
                            <input
                              class="baseNoPadding_656545-o_O-applyEllipsis_lspcml"
                              maxlength="255"
                              placeholder="a"
                              style="text-align: left; padding-left: 10px; padding-right: 32px;"
                              type="text"
                              value=""
                            />
                            <span
                              class="sigil_1ogu4v7"
                              style="right: 10px;"
                            >
                              <button
                                class="base_i7qsly-o_O-m_12juzcr-o_O-blank_1tdfl6w-o_O-responsive_zk4wup-o_O-button_yfi27k-o_O-none_v41adi"
                                type="button"
                              >
                                <span
                                  class="wrapper_17y0hv9 svgIcon"
                                  style="width: 12px; height: 12px; fill: unset;"
                                >
                                  <svg
                                    aria-hidden="true"
                                    aria-labelledby="title desc"
                                    class="svgCenter_1r81nla"
                                    color="unset"
                                    height="12"
                                    role="img"
                                    viewBox="0 0 64 64"
                                    width="12"
                                  >
                                    <title
                                      id="title"
                                    />
                                    <desc
                                      id="desc"
                                    >
                                      X
                                    </desc>
                                    <g
                                      fill-rule="nonzero"
                                    >
                                      <path
                                        d="M53.213 46.142L39.071 32l14.142-14.142a1 1 0 0 0 0-1.414l-5.657-5.657a1 1 0 0 0-1.414 0L32 24.929 17.858 10.787a1 1 0 0 0-1.414 0l-5.657 5.657a1 1 0 0 0 0 1.414L24.929 32 10.787 46.142a1 1 0 0 0 0 1.414l5.657 5.657a1 1 0 0 0 1.414 0L32 39.071l14.142 14.142a1 1 0 0 0 1.414 0l5.657-5.657a1 1 0 0 0 0-1.414z"
                                        fill-rule="evenodd"
                                        role="presentation"
                                      />
                                    </g>
                                  </svg>
                                </span>
                              </button>
                            </span>
                          </div>
                        </div>
                      </div>
                    }
                  />
                </Popper>
              </div>
            </Component>
          </div>
        </OnClickOutside>
      </OnClickOutside(OnClickOutside)>
    </Manager>
  </PopupWithClickAway>
</SearchableSelectInput>
`;
